INTEGRATION.md

```diff
@@ -0,0 +1,152 @@
+      F --> |no| J(try racing gateway*);
+      J --> |yes| K(test n-th gateway from list);
+      K --> KA(find the best** gateway);
+      J --> |no| L(fallback);
+      L --> |yes| M(dweb.link***);
+      L --> |no| N(error);
+```
+
+\* `try racing gateway` depends on the environment variable `IPFS_RACING_GATEWAY`. See below for details.
+
+\** `find the best* gateway`. The heuristics to find the best gateway are as follows. A racing gateway logic fires of a request to n-th gateways simultaneously (say 10 gateways out of a list of potentially 100's). Of those 10, the one that responds fastest is stored in a list. If this one is already responding within 50ms then this gateway is used. If this one is taking more then 200ms then the next batch of 10 gateways is probed. This flow continues till:
+1. a gateway with a response below 50ms has been found
+2. gateways have been probed for over 2 seconds (in which case it just stops and uses the fastest one).
+
+The result of this probe will be stored in `$CONFIG/ipfs/racing_gateway_response` as a single line being the gateway that responded fastest. In subsequent racing gateway requests this file will be read and used as starting point. If this gateway still responds within 50ms then no other gateways will be probed.
+
+\*** `dweb.link`, see the below `IPFS_FALLBACK_GATEWAY` for details.
+
+#### Environment variables
+
+The decision tree is influences by a couple environement variables.
+
+**`IPFS_RACING_GATEWAY`** When this environemnt variable isn't found (the default), racing gateways should be attempted upon reaching this point in the flow. When the environment variable exists it's value should be used. If it's value is `1` (or `true`) then racing gateways should be attempted. If this value is `0` (or `false`) then racing gateways are off. The control flow proceeds in the `no` branch.
+
+**`IPFS_FALLBACK_GATEWAY`** When this variable doesn't exist `dweb.link` will be used. When the variable does exist it's value will be used instead.
+
+### Gateway from command argument
+
+**This feature is optional and only for applications integrating IPFS support.**
+
+An application can opt to support a command line option to provide a gateway. If a user does provide this option then it should overrule any other gateway detection and be used as the gateway of choice. If implemented, it's recommended to go for either a `--gateway` or `--ipfs-gateway` argument. It depends very much on the application itself as to which option is most sensible.
+
+An example implementation that is doing this is ffmpeg with the ffplay utility. It allows the `-gateway` argument which by default is empty but can be set like: `-gateway http://127.0.0.1:8080` and would then be used to handle `ipfs://<cid>` or `ipns://<cid>`.
+
+### Gateway from IPFS_GATEWAY environment variable
+
+When there is no command line argument, the `IPFS_GATEWAY` environment is next. If it contains a value, it will be used as gateway.
```

Comment:
I highly prefer to not specify multiple here.
It would:
- be a pain parsing environments variables properly
- break ffmpeg (it expects 1)
- not match the naming (they indicate 1, not plural)

They also should be seen as "escape options" to get it working. The proper way would be through the gateway file.